app/db.ts

```diff
@@ -0,0 +1,374 @@
+		const query = dropTables(schema, dummyParent, undefined).reverse();
+		await this.query(query);
+	}
+}
+
+/**
+ * Gets name of id in child schema
+ * */
+function findid(schema: Schema) {
+	for (const field of schema) {
+		if (field.ty === DBTYPES.PK) {
+			return field.name;
+		}
+	}
+	console.error("A foreing key must be able to reference a primary key");
+	exit(1);
+}
+
+/**
+ * Returns [keys, values] of a schema
+ * */
+function keysAndValues(
+	schema: Schema,
+	run: Run,
+): [Array<string>, Array<string>] {
+	const keys = schema
+		.map((field) => {
+			switch (field.ty) {
+				case DBTYPES.PK:
+					return "";
+				default:
+					return field.name;
+			}
+		})
+		.filter((s) => s !== "");
+
+	const values = schema
+		.map((field) => {
+			switch (field.ty) {
+				case DBTYPES.PK:
+					return "";
+				case DBTYPES.TEXT:
+				case DBTYPES.INT:
+					return "'" + sanitize(run[field.runkey].toString()) + "'";
+				case DBTYPES.FK:
+					return `(SELECT id FROM ${field.table} WHERE ${createWhere(field.child, run)} LIMIT 1)`;
+			}
+		})
+		.filter((s) => s !== "");
+
+	return [keys, values];
+}
+
+/**
+ * Run a recursive command on all tables in the schema
+ * fact: Command special to the facttable
+ * child: Command special to all children
+ * */
+function traverseSchema<T>(
+	fact: (schema: Schema, parent: SchemaFK, val: T) => string[],
+	child: (schema: Schema, parent: SchemaFK, val: T) => string[],
+) {
+	return (schema: Schema, parent: SchemaFK, val: T) => {
+		let stmt: Array<string> = [];
+		// TRAVERSE CHILDREN
+		for (const field of schema) {
+			if (field.ty === DBTYPES.FK) {
+				stmt = [
+					...stmt,
+					...traverseSchema(child, child)(field.child, field, val),
+				];
+			}
+		}
+
+		// TRAVERSE SELF
+		return [...stmt, ...fact(schema, parent, val)];
+	};
+}
+
+/**
+ * Create the where clause of a child schema
+ * */
+function createWhere(schema: Schema, run: Run): string {
+	return schema
+		.map((field) => {
+			switch (field.ty) {
+				case DBTYPES.INT:
+					return `${field.name}='${run[field.runkey]}'`;
+				case DBTYPES.TEXT:
+					return `${field.name}='${sanitize(run[field.runkey] as string)}'`;
+				case DBTYPES.FK:
+					return `${field.name}=(SELECT id FROM ${field.table} WHERE ${createWhere(field.child, run)} LIMIT 1)`;
+				case DBTYPES.PK:
+					return "";
+			}
+		})
+		.filter((s) => s !== "")
+		.join(" AND ");
+}
+
+/**
+ * Insert a run in to the database
+ * */
+const insertRun = traverseSchema<Run>(
+	(schema, parent, run) => {
+		const [keys, values] = keysAndValues(schema, run);
+		return [
+			`INSERT INTO ${parent.table} (${keys.join(",")}) VALUES (${values.join(",")})`,
+		];
+	},
+	(schema, parent, run) => {
+		let stmt = "";
+
+		const [keys, values] = keysAndValues(schema, run);
+
+		stmt +=
+			`INSERT INTO ${parent.table} (${keys.join(",")}) SELECT ${values.join(",")} 
+			WHERE NOT EXISTS (SELECT 1 FROM ${parent.table} WHERE ${createWhere(parent.child, run)} LIMIT 1);`.replace(
+				/\s+/g,
```

Comment:
Hvorfor er det kun denne table input der fikser whitespaces, og ikke dem alle sammen?

Reply:
Fordi det er den eneste der er multiline, for læsbarheddens skyld